domain/application/state/containerimagemetadata.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	coreresource "github.com/juju/juju/core/resources"
+	"github.com/juju/juju/domain/application"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/internal/errors"
+)
+
+// GetContainerImageMetadata returns the container image metadata with the given
+// UUID. applicationerrors.ContainerImageMetadataNotFound is returned if the UUID is not in
+// the table
+func (s *State) GetContainerImageMetadata(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+) (application.ContainerImageMetadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Capture(err)
+	}
+
+	storageKey := containerImageMetadataStorageKey{StorageKey: resourceUUID.String()}
+	m := containerImageMetadata{}
+	stmt, err := s.Prepare(`
+SELECT &containerImageMetadata.*
+FROM   resource_container_image_metadata_store
+WHERE  storage_key = $containerImageMetadataStorageKey.storage_key
+`, storageKey, m)
+	if err != nil {
+		return application.ContainerImageMetadata{}, errors.Errorf(
+			"preparing select container image resource metadata statement: %w",
+			err,
+		)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, stmt, storageKey).Get(&m)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return applicationerrors.ContainerImageMetadataNotFound
+		}
+		return err
+	})
+	if err != nil {
+		return application.ContainerImageMetadata{}, err
+	}
+
+	return application.ContainerImageMetadata{
+		StorageKey:   m.StorageKey,
+		RegistryPath: m.RegistryPath,
+		Username:     m.UserName,
+		Password:     m.Password,
+	}, nil
+}
+
+// PutContainerImageMetadata stores container image metadata the database and
+// returns its UUID.
+func (s *State) PutContainerImageMetadata(
+	ctx context.Context,
+	resourceUUID coreresource.ID,
+	registryPath, userName, password string,
+) (string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Capture(err)
+	}
+
+	m := containerImageMetadata{
+		StorageKey:   resourceUUID.String(),
+		RegistryPath: registryPath,
+		UserName:     userName,
+		Password:     password,
+	}
+
+	stmt, err := s.Prepare(`
+INSERT INTO resource_container_image_metadata_store (*)
+VALUES ($containerImageMetadata.*) ON CONFLICT (storage_key) DO NOTHING
```

Comment:
Put usually implies replacing an existing entry. The same resource with a different path or credentials than prior would indicate to the caller that it was successfully applied, but this isn't the case.

Reply:
Do you mean that we should return an error if someone tries to put the same thing in twice? This isn't how the object store works, in that its a noop.

Reply:
I think what bother Joe is the usage of DO nothing where we could have use a DO UPDATE clause (and get ride of most of the code after which do the update in a second query).

From my perspective, as long as write query are costly, we should:

* If the expected behavior is to not update if there is already some data in the db => SELECT then INSERT (to avoid a write lock if already defined)
* if the expected behavior is to INSERT OR UPDATE... it would be a write anyway, so ON CONFLICT DO UPDATE may be what we want... But we may want to SELECT (read) then perform an INSERT or an UPDATE to keep the lock write as short as possible (not a db expert, it just feel that it may be better performance-wise, but we should back it with metrics).

I don't see any case were we actually want to use `ON CONFLICT DO NOTHING`, because it means that we acquire a write lock (which is a performance issue) for no reason.

Reply:
I've changed this to an upsert. I believe that implements the desired semantics.